### DISCUSSION TITLE
Publishing to message bus with HTTP trigger

### DISCUSSION THREAD

**Discussion by jrtitus** [edgexfoundry/community#145] [Q&A]
Is it still possible to publish to the message bus when using a Trigger type of `http`?

In v2 we could do this by creating a reading, explicitly pushing to core data, then setting the response data to return to the client:

```go
transforms.NewCoreDataObjectReading(
    config.GetValue(deviceNameKey).(string),
    config.GetValue(profileNameKey).(string),
    config.GetValue(resourceNameKey).(string)).PushToCoreData,
transforms.NewResponseData().SetResponseData,
```

However, in v3, we can only wrap the reading and set the response data:

```go
transforms.NewEventWrapperObjectReading(
    config.GetValue(profileNameKey).(string),
    config.GetValue(deviceNameKey).(string),
    config.GetValue(resourceNameKey).(string)).Wrap,
transforms.NewResponseData().SetResponseData,
```

According to the [documentation](https://docs.edgexfoundry.org/3.0/microservices/application/Triggers/#publishtopic), setting a `PublishTopic` and setting the response data should allow for publishing back to the message bus. Is this only applicable when the trigger type is `edgex-messagebus`?

For the `http` trigger, given a configuration such as:

```yml
Trigger:
  Type: http
  PublishTopic: events/device/device-service/device-profile/device-name/resource
```

would the publish to the bus still occur?

**Comment by lenny-goodell** [comment]
Setting response data with HTTP Trigger will result in the data being sent back as the HTTP response.

It is still possible with the use the Background Publisher from your Pipeline Function.
https://docs.edgexfoundry.org/3.0/microservices/application/ApplicationServiceAPI/#background-publisher-apis 
https://docs.edgexfoundry.org/3.0/microservices/application/AdvancedTopics/#background-publishing
It will use the configured `PublishTopic` or topic passed in code based on which API is used to create the Background Publisher.

This is made possible in 3.0 by the MessageBus enabled by default for all triggers (needed for Metrics)
Make sure you have not disabled the MessageBus in your configuration (not likely, but just check ;-) ):
https://github.com/edgexfoundry/app-service-configurable/blob/main/res/sample/configuration.yaml#L148

**Reply by jrtitus** [reply]
Hi @lenny-intel,

If I understand correctly, that means I'd need to do something like this to handle a one-shot publish with the `http` trigger?

```go
func (p *Publisher) PublishDataModel(ctx interfaces.AppFunctionContext, data interface{}) (bool, interface{}) {
	lc := ctx.LoggingClient()
	lc.Debug("Triggered PublishDataModel")

	bp, err := p.svc.AddBackgroundPublisher(1)
	if err != nil {
		lc.Debugf("failed to create publisher: %v", err.Error())
		return false, nil
	}

	if err := bp.Publish(ctx.ResponseData(), p.svc.BuildContext(uuid.NewString(), common.ContentTypeJSON)); err != nil {
		lc.Debugf("failed to publish data: %v", err.Error())
		return false, nil
	}

	return true, data
}
```

```go
publisher := functions.NewPublisher(service, database)

if err := service.SetDefaultFunctionsPipeline(
        // builds the response based on contents from an external database
	publisher.BuildDataModel,
        // wrap the response in EventWrapper
	transforms.NewEventWrapperObjectReading(
		config.GetValue(profileNameKey).(string),
		config.GetValue(deviceNameKey).(string),
		config.GetValue(resourceNameKey).(string)).Wrap,
        // use a background publisher to publish the response to the message bus
        // background publisher will automatically use Trigger.PublishTopic
	publisher.PublishDataModel,
	transforms.NewResponseData().SetResponseData,
); err != nil {
	lc.Errorf("SetDefaultFunctionsPipeline returned error: %s", err.Error())
	os.Exit(-1)
}
```

I have another application service configured to subscribe to the topic that this service is publishing onto - along the lines of:

```yml
Trigger:
    SubscribeTopics: events/+/+/+/device-name/#
```

but I still fail to retrieve any of the events in the other function pipeline.

The `MessageBus.Disabled` setting is undefined/false in each service configuration, and the `message-bus[service-name]` settings are configured in `EDGEX_ADD_KNOWN_SECRETS`.

---

I suppose I am mistaken because I tried to move the `AddBackgroundPublisher` to the `main` function of the application service and received this in the logs:

```
level=INFO ts=2023-07-17T19:59:45.14657245Z app=app-asset-manager source=triggerfactory.go:47 msg="HTTP trigger selected"
level=ERROR ts=2023-07-17T19:59:45.14657765Z app=app-asset-manager source=service.go:198 msg="background publishing not supported for services using HTTP trigger"
level=ERROR ts=2023-07-17T19:59:45.14658245Z app=app-asset-manager source=main.go:206 failedtoinitializeTrigger= msg="Run returned error: "
```

**Reply by lenny-goodell** [reply]
Looks like you have exposed a bug with background publishing when using non-MessageBus trigger. The handling publishing is only currently implemented in the MessageBus trigger itself, so the Events you publish are still sitting in the queue. I will create an issue for this

The only other option is to use the EventClient() from the `ctx` parameter to recreate `PushToCore` by calling the `.Add()` function in your PublishDataModel function. The Event.Wrap function returns an `AddEventRequest`, so you the data passed to your function is what you need to pass to `.Add()`

**Reply by jrtitus** [reply]
> The only other option is to use the EventClient() from the `ctx` parameter to recreate `PushToCore` by calling the `.Add()` function in your PublishDataModel function. The Event.Wrap function returns an `AddEventRequest`, so you the data passed to your function is what you need to pass to `.Add()`

That looks like it resolves the problem I was having - the other app service is picking up the event successfully now if I do this:

```go
func (p *Publisher) PublishDataModel(ctx interfaces.AppFunctionContext, data interface{}) (bool, interface{}) {
	lc := ctx.LoggingClient()
	lc.Debug("Triggered PublishDataModel")

	c := context.Background()
	_, err := ctx.EventClient().Add(c, icommon.ServiceName, data.(requests.AddEventRequest))
	if err != nil {
		lc.Debugf("failed to publish data: %v", err)
		return false, nil
	}

	return true, data
}
```
